airflow/api/common/experimental/mark_tasks.py

```diff
@@ -0,0 +1,187 @@
+    :param execution_dates: list of execution dates to evaluate
+    :param state: the state to set the dag run to
+    :param run_id_template:the template for run id to be with the execution date
+    :return: newly created and existing dag runs for the execution dates supplied
+    """
+    # find out if we need to create any dag runs
+    drs = DagRun.find(dag_id=dag.dag_id, execution_date=execution_dates)
+    dates_to_create = list(set(execution_dates) - set([dr.execution_date for dr in drs]))
+
+    for date in dates_to_create:
+        dr = dag.create_dagrun(
+            run_id=run_id_template.format(date.isoformat()),
+            execution_date=date,
+            start_date=datetime.datetime.now(),
+            external_trigger=False,
+            state=state,
+        )
+        drs.append(dr)
+
+    return drs
+
+
+def set_state(task, execution_date, upstream=False, downstream=False,
+              future=False, past=False, state=State.SUCCESS, commit=False):
+    """
+    Set the state of a task instance and if needed its relatives. Can set state
+    for future tasks (calculated from execution_date) and retroactively
+    for past tasks. Will verify integrity of past dag runs in order to create
+    tasks that did not exist. It will not create dag runs that are missing
+    on the schedule (but it will as for subdag dag runs if needed).
+    :param task: the task from which to work. task.task.dag needs to be set
+    :param execution_date: the execution date from which to start looking
+    :param upstream: Mark all parents (upstream tasks)
+    :param downstream: Mark all siblings (downstream tasks) of task_id, including SubDags
+    :param future: Mark all future tasks on the interval of the dag up until
+        last execution date.
+    :param past: Retroactively mark all tasks starting from start_date of the DAG
+    :param state: State to which the tasks need to be set
+    :param commit: Commit tasks to be altered to the database
+    :return: list of tasks that have been created and updated
+    """
+    assert isinstance(execution_date, datetime.datetime)
+
+    # microseconds are supported by the database, but is not handled
+    # correctly by airflow on e.g. the filesystem and in other places
+    execution_date = execution_date.replace(microsecond=0)
+
+    assert task.dag is not None
+    dag = task.dag
+
+    latest_execution_date = dag.latest_execution_date
+    assert latest_execution_date is not None
+
+    # determine date range of dag runs and tasks to consider
+    end_date = latest_execution_date if future else execution_date
+
+    if 'start_date' in dag.default_args:
+        start_date = dag.default_args['start_date']
+    elif dag.start_date:
+        start_date = dag.start_date
+    else:
+        start_date = execution_date
+
+    start_date = execution_date if not past else start_date
+
+    if dag.schedule_interval == '@once':
+        dates = [start_date]
```

Comment:
not a big deal for us since we don't use `@once` but if we manually schedule more dags after the first dag run changing task state or clearing tasks it will only change the very first dag run